python/pyspark/__init__.py

```diff
@@ -63,44 +64,56 @@
 from pyspark.version import __version__
 from pyspark._globals import _NoValue  # noqa: F401
 
+T = TypeVar("T")
+F = TypeVar("F", bound=Callable)
 
-def since(version):
+
+def since(version: Union[str, float]) -> Callable[[T], T]:
     """
     A decorator that annotates a function to append the version of Spark the function was added.
     """
     import re
 
     indent_p = re.compile(r"\n( +)")
 
-    def deco(f):
-        indents = indent_p.findall(f.__doc__)
+    def deco(f: T) -> T:
```

Comment:
Shouldn't this be `F -> F`?

Reply:
I think we expect the output is ```Callable[[T], T]``` so deco should be ```T -> ```. Did I misunderstand?

Reply:
The `deco` will take a function to be decorated and return the decorated function.
So it should be `def deco(f: F) -> F:` and the return type of `since` should be `Callable[[F], F]`.

Reply:
Thank you! updated.